src/Bicep.Core/Semantics/Namespaces/SystemNamespaceType.cs

```diff
@@ -281,7 +283,16 @@ private static IEnumerable<FunctionOverload> GetSystemOverloads(IFeatureProvider
 
             yield return new FunctionOverloadBuilder("first")
                 // TODO even with non-literal types, some type arithmetic could be performed
-                .WithReturnResultBuilder(TryDeriveLiteralReturnType("first", LanguageConstants.Any), LanguageConstants.Any)
+                .WithReturnResultBuilder((binder, fileResolver, diagnostics, arguments, argumentTypes) =>
+                {
+                    return argumentTypes[0] switch
+                    {
+                        TypedArrayType typedArrayType => new(typedArrayType.Item.Type),
+                        TupleType tupleType when tupleType.Items.Length == 0 => new(LanguageConstants.Null),
+                        _ => TryDeriveLiteralReturnType("first", LanguageConstants.Any)
+                                .Invoke(binder, fileResolver, diagnostics, arguments, argumentTypes)
+                    };
+                }, TypeHelper.CreateTypeUnion(LanguageConstants.Object, LanguageConstants.Any, LanguageConstants.Null))
```

Comment:
```suggestion
                }, LanguageConstants.Null)
```

[Nit] A type union of `LanguageConstants.Any` and anything else will be `LanguageConstants.Any`, so there's no need to create the union here.